src/realm/sync/noinst/migration_store.cpp

```diff
@@ -68,61 +62,84 @@ MigrationStore::MigrationStore(DBRef db,
+    if (auto migration_table = tr->get_table(m_migration_table); migration_table->is_empty()) {
+        tr->promote_to_write();
+        auto migration_store_obj = migration_table->create_object();
+        migration_store_obj.set(m_migration_state, int64_t(MigrationState::NotStarted));
+        m_state = MigrationState::NotStarted;
+        tr->commit();
+    }
+    else {
+        auto migration_store_obj = migration_table->get_object(0);
+        m_state = static_cast<MigrationState>(migration_store_obj.get<int64_t>(m_migration_state));
+        m_query_string = migration_store_obj.get<String>(m_migration_query_str);
+    }
 }
 
 std::shared_ptr<realm::SyncConfig> MigrationStore::convert_sync_config(std::shared_ptr<realm::SyncConfig> config)
 {
     REALM_ASSERT(config);
 
     std::lock_guard lock{m_mutex};
-    if (config->flx_sync_requested) {
-        cancel_migration();
-        return config;
+    if (m_state == MigrationState::Completed && config->flx_sync_requested) {
+        // Migration state is no longer needed - natively using flexible sync
+        reset_migration_info();
+        return config; // return original sync config
     }
     if (m_state == MigrationState::NotStarted) {
+        // Using PBS and not migrated, return original sync config
         return config;
     }
 
+    // Using PBS and migrated, update copy of original sync config to FLX and return
     auto flx_config = std::make_shared<realm::SyncConfig>(*config); // deep copy
-    flx_config->partition_value = "";
+    flx_config->partition_value = {};
     flx_config->flx_sync_requested = true;
 
     return flx_config;
 }
 
-void MigrationStore::migrate_to_flx(std::string rql_query_string)
+void MigrationStore::migrate_to_flx(std::string_view rql_query_string)
 {
     REALM_ASSERT(!rql_query_string.empty());
 
-    std::unique_lock lock{m_mutex};
-    m_query_string = rql_query_string;
-    m_state = MigrationState::Completed;
-    lock.unlock();
-
-    auto tr = m_db->start_write();
-    auto migration_store_obj = tr->get_table(m_migration_table)->get_object(0);
-    migration_store_obj.set(m_migration_query_str, m_query_string);
-    migration_store_obj.set(m_migration_state, int64_t(m_state));
-    migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
-    tr->commit();
+    {
+        std::unique_lock lock{m_mutex};
+        m_query_string = rql_query_string;
+        m_state = MigrationState::Completed;
+
+        auto tr = m_db->start_write();
+        auto migration_store_obj = tr->get_table(m_migration_table)->get_object(0);
+        migration_store_obj.set(m_migration_query_str, m_query_string);
+        migration_store_obj.set(m_migration_state, int64_t(m_state));
+        migration_store_obj.set(m_migration_completed_at, Timestamp{std::chrono::system_clock::now()});
+        tr->commit();
+    }
 
-    m_on_migration_state_changed(m_state);
+    m_on_migration_state_changed(MigrationState::Completed);
```

Comment:
I feel like there's some intermediate state here where we haven't actually migrated, we've just created the migration store object and set its properties and the actual process of migration is actually about to begin rather than be completed.

Reply:
We're both on the fence about whether or not the third state is needed. For now we're starting with two and will add the intermediate state while the client reset/change of query process is underway.